### PR TITLE
Fix various potential XSS vulnerabilities

### PR DESCRIPTION
**Pull request by cpeel**
To prevent XSS from invalid input, correctly escape raised exception messages.

Thanks to M. Dipu for finding and reporting this problem.
OBB-1281669

Note: waiting to hear back from M. Dipu for how they want to be recognized for finding this flaw. This is hotfixed on PROD already.

### CODE REVIEW

**Review by 70ray — APPROVED** [review]

**Comment by cpeel** [comment]
I've run a vulnerability scanner against TEST and it found a few more things that I'm addressing for logged-in users. The new updates include changes to how we redirect users when setting the language cookie and redirecting to pages requiring login.

This is in the [fix-xss](https://www.pgdp.org/~cpeel/c.branch/fix-xss/default.php) sandbox for folks that want to play with it.

**Comment by cpeel** [comment]
After more testing with a 3rd party tool on the TEST site, I've addressed a few more problems. Functionality that is worth (re)testing:
* Login / logout
* Project Search -- including changing the columns via the Configure Result link
* While not logged in, set the language and be redirected to the page you were on
* When logging in from a page not on the activity hub, being returned to the page you are on
* Subscribing to project events
* Viewing past news
* Confirming you can proof a new page

Virtually all of the changes should be entirely behind-the-scenes and not visible to the user.

**Comment by srjfoo** [comment]
> After more testing with a 3rd party tool on the TEST site, I've addressed a few more problems. Functionality that is worth (re)testing:
> 
>     * Login / logout

Looks good

>     * Project Search -- including changing the columns via the Configure Result link

Looks good.

>     * While not logged in, set the language and be redirected to the page you were on

From the search results page: Logged out, and tried logging back in in all three specific languages (Icelandic, French and English), and logged back in. In all three cases, I was redirected to the Activity Hub.

>     * When logging in from a page not on the activity hub, being returned to the page you are on

See above.

>     * Subscribing to project events

Looks good.

>     * Viewing past news

I thought I couldn't see the "See all random news items" link at all, but eventually figured out that it was because most of the pages didn't have random news items. I have added random news items for most pages.

>     * Confirming you can proof a new page

Looks good.

**Comment by cpeel** [comment]
> > ```
> > * While not logged in, set the language and be redirected to the page you were on
> > ```
> 
> From the search results page: Logged out, and tried logging back in in all three specific languages (Icelandic, French and English), and logged back in. In all three cases, I was redirected to the Activity Hub.

I should have been more specific on this and the next one. While not logged in, go to the Register page. Change the site language in the sidebar and you should return to the Register page.

> > ```
> > * When logging in from a page not on the activity hub, being returned to the page you are on
> > ```
> 
> See above.

While logged in, go to a page (User Preferences, Project Search results, etc). Copy the URL. Log out. Paste in the URL as if you were clicking a link or returning from a bookmark. You'll be redirected to a "login required" page. Log in. You should be returned to the page you started with.

**Comment by srjfoo** [comment]
> > > * While not logged in, set the language and be redirected to the page you were on
> > 
> > From the search results page: Logged out, and tried logging back in in all three specific languages (Icelandic, French and English), and logged back in. In all three cases, I was redirected to the Activity Hub.
> 
> I should have been more specific on this and the next one. While not logged in, go to the Register page. Change the site language in the sidebar and you should return to the Register page.

Thanks for the clarification. Yes, changing languages while logged out and on the registration page shows the translated registration page

> > > * When logging in from a page not on the activity hub, being returned to the page you are on
>>
> > See above.
>
> While logged in, go to a page (User Preferences, Project Search results, etc). Copy the URL. Log out. Paste in the URL as if you were clicking a link or returning from a bookmark. You'll be redirected to a "login required" page. Log in. You should be returned to the page you started with.

Yep. Works as described.

File: project.php [hunk 1]
```diff
@@ -1163,7 +1163,6 @@ function do_event_subscriptions()
     $url = "$code_url/tools/set_project_event_subs.php";
     echo "<form method='post' action='$url'>\n";
     echo "<input type='hidden' name='projectid' value='$projectid'>\n";
-    echo "<input type='hidden' name='return_uri' value='" . urlencode($_SERVER['REQUEST_URI']) . "#event_subscriptions'>\n";
```
[the remark this answers is not in the document]
Comment (cpeel) [reply project.php:1166]: No it isn't, but this line was removed so that's OK 😁

**Pull request merged**